### PR TITLE
New Resource and Data Source: aws_eks_cluster

### LINKED ISSUE #2473: Amazon EKS Resource (Elastic Container Service for Kubernetes)
Amazon Elastic Container Service for Kubernetes has been released. Supporting it will be super cool.

https://aws.amazon.com/blogs/aws/amazon-elastic-container-service-for-kubernetes/

### PR DESCRIPTION
**Pull request by bflad**
Closes #2473 

Changes proposed in this pull request:

* New Resource: `aws_eks_cluster`
* New Data Source: `aws_eks_cluster`
* New Guide: EKS Getting Started

Output from acceptance testing:

```
Pending in TeamCity
```

### CODE REVIEW

File: aws/data_source_aws_eks_cluster.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"data": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+					},
+				},
+			},
+			"created_at": {
```
Comment (radeksimko) [inline aws/data_source_aws_eks_cluster.go]: Is this field actually useful for anyone/anything? Not a big deal if not, but I thought historically we avoided encoding fields which aren't useful from user's perspective to the schema.

**Comment by mechastorm** [comment]
Wow this is the fastest I've seen an AWS service integrated into the terraform provider. Great work and thanks for the prompt update!

File: aws/resource_aws_eks_cluster.go [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+			if isAWSErr(err, eks.ErrCodeInvalidParameterException, "does not exist") {
+				return resource.RetryableError(err)
+			}
+			if isAWSErr(err, eks.ErrCodeInvalidParameterException, "Role could not be assumed because the trusted entity is not correct") {
+				return resource.RetryableError(err)
+			}
+			// InvalidParameterException: The provided role doesn't have the Amazon EKS Managed Policies associated with it. Please ensure the following policies [arn:aws:iam::aws:policy/AmazonEKSClusterPolicy, arn:aws:iam::aws:policy/AmazonEKSServicePolicy] are attached
+			if isAWSErr(err, eks.ErrCodeInvalidParameterException, "The provided role doesn't have the Amazon EKS Managed Policies associated with it") {
+				return resource.RetryableError(err)
+			}
```
Comment (radeksimko) [inline aws/resource_aws_eks_cluster.go]: 😭

File: aws/resource_aws_eks_cluster.go [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+	}
+
+	_, err := conn.DeleteCluster(input)
+	if err != nil {
+		if isAWSErr(err, eks.ErrCodeResourceNotFoundException, "") {
+			return nil
+		}
+		// Sometimes the EKS API returns the ResourceNotFound error in this form:
+		// ClientException: No cluster found for name: tf-acc-test-0o1f8
+		if isAWSErr(err, eks.ErrCodeClientException, "No cluster found for name:") {
```
Comment (radeksimko) [inline aws/resource_aws_eks_cluster.go]: 😭

File: aws/resource_aws_eks_cluster_test.go [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+			break
+		}
+
+		input.NextToken = out.NextToken
+	}
+
+	return nil
+}
+
+func TestAccAWSEksCluster_Basic(t *testing.T) {
```
Comment (radeksimko) [inline aws/resource_aws_eks_cluster_test.go]: Nitpick, but AFAIK all our tests use `_basic`, not `_Basic`.

File: aws/resource_aws_eks_cluster_test.go [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+
+func testAccCheckAWSEksClusterDestroy(s *terraform.State) error {
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "aws_eks_cluster" {
+			continue
+		}
+
+		conn := testAccProvider.Meta().(*AWSClient).eksconn
+
+		// Handle eventual consistency
```
Comment (radeksimko) [inline aws/resource_aws_eks_cluster_test.go]: 😭

**Comment by bflad** [comment]
FYI I'm just waiting for the official EKS User Guide to be published so I can verify some pieces of the getting started guide. Some pieces might have changed between preview and general availability.

File: website/docs/guides/eks-getting-started.html.md [hunk 6]
```diff
@@ -0,0 +1,594 @@
...
+  Using Terraform to configure AWS EKS.
+---
+
+# Getting Started with AWS EKS
+
+The Amazon Web Services EKS service allows for simplified management of
+[Kubernetes](https://kubernetes.io/) servers. While the service itself is
+quite simple from an operator perspective, understanding how it interconnects
+with other pieces of the AWS service universe and how to configure local
+Kubernetes clients to managed clusters can be daunting at first.
```
Comment (radeksimko) [inline website/docs/guides/eks-getting-started.html.md]: While I might agree with your statement I'm not sure we want to put ourselves in the position of someone who can say what 3rd party service/product is or isn't _daunting_ in our docs generally. 🤔

File: website/docs/guides/eks-getting-started.html.md [hunk 7]
```diff
@@ -0,0 +1,594 @@
...
+AWS services as well as allow inbound networking communication from your
+local workstation (if desired) and worker nodes.
+
+#### EKS Master Cluster IAM Role
+
+The below is an example IAM role and policy to allow the EKS service to
+manage or retrieve data from other AWS services. It is also possible to create
+these policies with the [`aws_iam_policy_document` data source](/docs/providers/aws/d/iam_policy_document.html)
+
+For the latest required policy, see the EKS User Guide.
```
Comment (radeksimko) [inline website/docs/guides/eks-getting-started.html.md]: Nitpick, but link to the EKS User Guide may be useful here.

File: website/docs/guides/eks-getting-started.html.md [hunk 8]
```diff
@@ -0,0 +1,594 @@
...
+generates a sample `kubectl` configuration to connect to your cluster.
+
+You can verify cluster access via `kubectl version` displaying server version
+information in addition to local client version information.
+
+```hcl
+locals {
+  kubeconfig = <<KUBECONFIG
+
+
```
Comment (radeksimko) [inline website/docs/guides/eks-getting-started.html.md]: Are the two empty lines desired here?

File: website/docs/guides/eks-getting-started.html.md [hunk 9]
```diff
@@ -0,0 +1,594 @@
...
+finally a configuration that allows automatic scaling of worker nodes.
+
+#### Worker Node IAM Role and Instance Profile
+
+The below is an example IAM role and policy to allow the worker nodes to
+manage or retrieve data from other AWS services. It is used by Kubernetes
+to allow worker nodes to join the cluster. It is also possible to create
+these policies with the [`aws_iam_policy_document` data source](/docs/providers/aws/d/iam_policy_document.html)
+
+For the latest required policy, see the EKS User Guide.
```
Comment (radeksimko) [inline website/docs/guides/eks-getting-started.html.md]: Likewise, link to the user guide would be nice to have.

File: website/docs/guides/eks-getting-started.html.md [hunk 10]
```diff
@@ -0,0 +1,594 @@
...
+used in conjunction with AutoScaling policies (not implemented here).
+
+First, let us create a data source to fetch the latest Amazon Machine Image
+(AMI) that Amazon provides with an EKS compatible Kubernetes baked in.
+
+```hcl
+data "aws_ami" "eks-worker" {
+  filter {
+    name   = "name"
+    values = ["eks-worker-*"]
```
Comment (radeksimko) [inline website/docs/guides/eks-getting-started.html.md]: Just curious - this is version agnostic intentionally? i.e. is the data source always going to fetch a valid AMI regardless of the K8S version (currently 1.10)?

File: website/docs/guides/eks-getting-started.html.md [hunk 11]
```diff
@@ -0,0 +1,594 @@
...
+
+#### Required Kubernetes Configuration to Join Worker Nodes
+
+The EKS service does not provide a cluster-level API parameter or resource to
+automatically configure the underlying Kubernetes cluster to allow worker nodes
+to join the cluster via AWS IAM role authentication.
+
+While managing the underlying Kubernetes cluster configuration is beyond the
+scope of this guide, we provide an example of how to apply the required
+Kubernetes `ConfigMap` via `kubectl` below for completeness.
```
Comment (radeksimko) [inline website/docs/guides/eks-getting-started.html.md]: A link to Config Map would be nice to have here.
https://cloud.google.com/kubernetes-engine/docs/concepts/configmap

File: website/docs/guides/eks-getting-started.html.md (hunk 6 above)
Comment (bflad) [reply website/docs/guides/eks-getting-started.html.md]: 👍  I'll update the verbiage here to read `can be helpful.`

File: website/docs/guides/eks-getting-started.html.md (hunk 7 above)
Comment (bflad) [reply website/docs/guides/eks-getting-started.html.md]: Ha! Great catch. I was waiting for the official URL. I'll add it.

File: website/docs/guides/eks-getting-started.html.md (hunk 8 above)
Comment (bflad) [reply website/docs/guides/eks-getting-started.html.md]: When I was developing the original demo configuration, the extra lines were helpful for distinguishing the "config" output from the "TF" output in the terminal. I don't think they'll hurt anything being in here still.

File: website/docs/guides/eks-getting-started.html.md (hunk 10 above)
Comment (bflad) [reply website/docs/guides/eks-getting-started.html.md]: I think we can address this at a later time by updating the various pieces that might be version specific. 👍  Currently the guide is version agnostic.

File: website/docs/guides/eks-getting-started.html.md (hunk 11 above)
Comment (errm) [reply website/docs/guides/eks-getting-started.html.md]: http://kubernetes.io/docs/user-guide/configmap/ would be better IMO

Comment (bflad) [reply website/docs/guides/eks-getting-started.html.md]: Awesome! Will add.

Comment (bflad) [reply website/docs/guides/eks-getting-started.html.md]: FYI we'll update the link to the one from @errm post-merge 👍

**Pull request merged**